optimum/amd/brevitas/quantizer.py

```diff
@@ -0,0 +1,319 @@
+
+        # Because accelerate is not compatible with FX, we keep two versions of the Model
+        # one with FX-traced, the other one not.
+        # Since weights are shared across the two, we can apply weight/activation equalization
+        # by using one representation or the other based on needs.
+
+        if quantization_config.apply_weight_equalization:
+            logger.info("Applying weight equalization...")
+            apply_weight_equalization(model)
+            logger.info("Weight equalization applied.")
+
+        if quantization_config.activations_equalization is not None:
+            logger.info(
+                f"Applying Activation Equalization {quantization_config.activations_equalization} (SmoothQuant)..."
+            )
+            apply_act_equalization(model, quantization_config.activations_equalization, calibration_dataset)
+            logger.info("Activation equalization applied.")
+
+        # We do not quantize embedding and last fully connected layer
+        model = quantize_model(
+            model,
+            dtype=dtype,
+            weight_quant_format="int",
+            weight_quant_type="sym" if quantization_config.weights_symmetric else "asym",
+            weight_bit_width=quantization_config.weights_bitwidth,
+            weight_param_method=quantization_config.weights_param_method,
+            weight_scale_precision=quantization_config.scale_precision,
+            weight_quant_granularity=quantization_config.weights_quant_granularity,
+            weight_group_size=quantization_config.weights_group_size,
+            quantize_weight_zero_point=quantization_config.quantize_zero_point,
+            input_bit_width=quantization_config.activations_bitwidth,
+            input_quant_type="sym" if quantization_config.activations_symmetric else "asym",
+            input_quant_format="int",
+            input_param_method=quantization_config.activations_param_method,
+            input_scale_precision=quantization_config.scale_precision,
+            input_scale_type="static" if quantization_config.is_static else "dynamic",
+            input_quant_granularity=quantization_config.activations_quant_granularity,
+            input_group_size=quantization_config.activations_group_size,
+            quantize_input_zero_point=quantization_config.quantize_zero_point,
+        )
+
+        # Perform a single inference pass to generate the correct state_dict. This is necessary as Brevitas has some magic where
+        # a first forward pass need to be called before quantizing a model:
+        # https://github.com/Xilinx/brevitas/blob/84f42259ec869eb151af4cb8a8b23ad925f493db/src/brevitas/core/scaling/standalone.py#L205-L217
+        with torch.no_grad():
+            if calibration_dataset is not None:
+                model(**calibration_dataset[0])
+            else:
+                # TODO: clean that with a proper dummy input.
+                model(input_ids=torch.tensor([[1]], dtype=torch.int64))
```

Comment:
This will break in case the model is FX traced

Reply:
isn't a calibration dataset always required when using fx?

Reply:
Weight equalization can be done with FX without a calibration dataset

Reply:
should be fixed in https://github.com/huggingface/optimum-amd/pull/54/commits/30ae43c9edf790d5432336d3184836f69159bfcf

Reply:
It is still not fixed, since the sample is created but not used.